### PR TITLE
Add build-level retry

### PR DESCRIPTION
**Pull request by blindpirate**
### Context 

We've been annoyed by non-test flakiness for a long time. This PRs adds the build level retry support for CI builds. For functional tests, if `GRADLE_RUNNER` step fails, there will be 3 extra steps running:

- KILL_ALL_GRADLE_PROCESSES: kill all Gradle JVMs on the build agent, including all daemons.
- CLEAN_UP_GIT_UNTRACKED_FILES_AND_DIRECTORIES: remove all on tracked files in the checkout directory.
- GRADLE_RETRY_RUNNER: rerun the failed builds.

In short, we retry the build step in a fresh checkout directory with a fresh environment. We hope this could fix most flakiness on CI. For now, this is only enabled for pre-tested commit builds.

### CODE REVIEW

File: .teamcity/src/main/kotlin/common/extensions.kt [hunk 1]
```diff
@@ -191,6 +192,12 @@ fun BuildSteps.checkCleanM2AndAndroidUserHome(os: Os = Os.LINUX, buildType: Buil
     }
 }
 
+fun BuildStep.onlyRunOnPreTestedCommitBuildBranch(buildType: BuildType? = null) {
```
Comment (wolfs) [inline .teamcity/src/main/kotlin/common/extensions.kt:195]: ❓ Why do you need the `buildType` as an argument here?

File: build-logic/cleanup/src/main/java/gradlebuild/cleanup/services/KillLeakingJavaProcesses.java [hunk 2]
```diff
@@ -42,42 +42,87 @@
  * Usage: java build-logic/cleanup/src/main/java/gradlebuild/cleanup/services/KillLeakingJavaProcesses.java
  */
 public class KillLeakingJavaProcesses {
+    enum ExecutionMode {
+        /**
+         * Run at the beginning of each build. Kill potentially leaked processes in previos builds.
```
Comment (wolfs) [inline build-logic/cleanup/src/main/java/gradlebuild/cleanup/services/KillLeakingJavaProcesses.java:47]:
```suggestion
         * Run at the beginning of each build. Kill potentially leaked processes in previous builds.
```

💅  typo

File: build-logic/cleanup/src/main/java/gradlebuild/cleanup/services/KillLeakingJavaProcesses.java [hunk 3]
```diff
@@ -42,42 +42,87 @@
...
-        return "(?i)[/\\\\]" + "(" + javaExecutable + ".+?" + "(?:" +
-            classPathPattern1 + "|" +
-            classPathPattern2 + "|" +
-            classPathPattern3 + "|" +
-            kotlinCompilerDaemonPattern + "|" +
-            javaProcessStackTracesMonitorPattern +
-            ").+)";
+        String classPathPattern1 = "(?:-cp.+" + "(" + quotedRootProjectDir + "|build\\\\tmp\\\\performance-test-files)" + ".+?" + GRADLE_MAIN_CLASS_PATTERN_STR + ")";
+        String classPathPattern2 = "(?:-classpath.+" + quotedRootProjectDir + ".+?" + Pattern.quote("\\build\\") + ".+?" + GRADLE_MAIN_CLASS_PATTERN_STR + ")";
+        String classPathPattern3 = "(?:-classpath.+" + quotedRootProjectDir + ".+?" + PLAY_SERVER_PATTERN_STR + ")";
```
Comment (wolfs) [inline build-logic/cleanup/src/main/java/gradlebuild/cleanup/services/KillLeakingJavaProcesses.java:87]: ❓ Do you have better names than `classPathPattern{1-3}`?

**Review by wolfs — APPROVED** [review]
Nice work! I left some smaller comments. Did you try this out in the experimental pipeline already?
